[JENKINS-19022] rebuild BuildData objects instead of storing them in the build.xml

jacob-keller opened the pull request:
The BuildData objects are serialized into the build.xml for every build.
This is normally not a problem, as their contents are small. However,
for certain types of builds (gerrit-triggered projects, really any git
build which uses a large number of refs) the BuildData objects can
become enormous. Worse, in those cases the data it saves is nearly
worthless.

Unfortunately, complete removal of the BuildData objects is not
possible, as we rely on this data to correctly predict change logs, and
to determine what revisions to build. Disentangling all of these
operations from the BuildData is tricky, if not impossible. This is
especially so for extension points such as the BuildChoosers.

Instead, lets opt for a different approach: generate the BuildData
instead of serializing the buildsByBranchName map to the build.xml for
every single build.

Do this by implementing a simpler BuildDetails structure, which contains
most of the same information, but does not contain the
buildsByBranchName map.

Stop storing the BuildData object into the build Actions. Instead,
search for BuildDetails and regenerate the BuildData structure from
this. As a fallback, allow finding old BuildData actions and using
those, if they still exist.

Jenkins Administrators may which to purge this data from their build.xml
files if they do not need the historical data, but for now, keep the
plugin as backwards compatible as possible.

With this change, we now only store a single Build object per build
instead of storing the entire branch map, which should significantly
reduce build.xml bloat.

If the time cost of rebuilding the BuildData structure is too high,
(such as when there are a large number of historical builds), we could
add some sort of limit on the number of past builds to check. However,
this would require some thought, as it would not produce exact replicas
of the current BuildData objects.

See [JENKINS-19022](https://issues.jenkins-ci.org/browse/JENKINS-19022) and  [JENKINS-47789](https://issues.jenkins-ci.org/browse/JENKINS-47789) for more details on the issues
the BuildData objects can cause.

Signed-off-by: Jacob Keller <jacob.e.keller@intel.com>

TODO:
- [x] Determine impact with a large number of jobs
- [x] Ensure we have enough test cases to cover things
- [x] Determine proper documentation required?
- [ ] Maybe think about providing a method of removing the old data? Or maybe we automate it within the plugin?? 
- [ ] Maybe think about adding some limiter for the jobs to dig backwards? i.e. only dig back a few hundred jobs maximum?

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
(I'm seeing some weird timeout issues when I run the tests... It may be because I'm behind a proxy?)
- [ ] I have added documentation as necessary
(We definitely want to document this as a change since others may be reading the build data)
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
(Not yet, it's on my list of TODO)
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [x] Bug fix (breaking change which fixes an issue)

I'm not sure if this constitutes a breaking change or not. We no longer export BuildData, Instead exporting BuildDetails in the ActionableItems list.. This could(?) impact consumers of the data? but they can learn to rebuild the data just like we do in core??

## Further comments

This approach was the easiest I've found so far. If we actually try to remove BuildData, it turns out a lot of the core code depends on being able to look some of the things up. It's possible we really want to limit the number of builds we need to search for, but I don't actually have data on how slow it is right now.

I welcome thoughts and criticism of this approach!

tsuna commented:
How well is this going to scale for jobs that have >>10k of prior builds?

MarkEWaite commented:
I'm not sure it could scale any worse than the current implementation.  I haven't performed any specific benchmarks, but have bug reports of bad results as many builds are retained.

jacob-keller commented:
> I'm not sure it could scale any worse than the current implementation. I haven't performed any specific benchmarks, but have bug reports of bad results as many builds are retained.

It depends. If the builds have only one branch, then it will scale VERY poorly compared to the current solution. (ie: currently the map will be one branch per build, and would be very small to serialize from the builds). However, I don't know if that scaling will be poor in *practice*. I think we could get away with some sort of depth limit, where we say "the map will only be accurate back say 100 builds" rather than being accurate out to the full history. For how we use it, I think this trade off would be acceptable, assuming that digging open 10k builds actually is a problem.

For builds with many branches (one per build) such as gerrit change-id based builds, this will be pretty much the same as current (the map will contain one entry per build), unless loading each one takes longer?

I'd like to get actual results for this soon but I'm not sure when or how best to test that.

jacob-keller commented:
I'll take a look at fixing those find bug warnings later this evening.

jacob-keller commented:
Huh, well I fixed the first set of test failures, and that broke something else... I'm still not able to get the tests to run locally :(

jacob-keller commented:
Finally figured out what I was doing wrong when constructing the build data objects. "getBuildData()" expects to be able to dig farther back into history if the top build doesn't contain any build information, and i wasn't allowing that.

I also finally got my tests to resolve properly on my local system!

MarkEWaite commented:
@jacob-keller I don't understand why the tests don't run locally for you.  I'm running the tests locally on 

* Windows 7 
* Windows 10
* CentOS 6
* CentOS 7
* Debian 8
* Debian 9
* Ubuntu 14
* Ubuntu 16
* FreeBSD 10

Sometimes the tests are flaky on one or more of those platforms, but they definitely run.  Can you share more information about the failure mode of the tests when you run them in your environment?

jacob-keller commented:
> Sometimes the tests are flaky on one or more of those platforms, but they definitely run. Can you share more information about the failure mode of the tests when you run them in your environment?

I should clarify: they run, but several tests fail.

I'll attach the test output of my mvn test run here as git-plugin-test-output.txt

[git-plugin-test-output.txt](https://github.com/jenkinsci/git-plugin/files/1831019/git-plugin-test-output.txt)

I *suspect* this is because it tries to contact remote repositories and fails to use the proxy settings when talking to them.

jacob-keller commented:
Additionally, due to 2minute timeouts, this causes the test suite to take an extremely long time to finish.

MarkEWaite commented:
@jacob-keller I merged your changes with the contents of the require-jdk-8 branch and the daniel-beck-make-matrix-optional branch and installed it into my Jenkins environment for testing.

Unfortunately, one of the earliest jobs I checked reported a new null pointer exception.  I see it in many of my jobs.  Before I spend time describing the detailed steps to reporduce the problem, could you try your build of the plugin in an environment that uses a Pipeline shared library.

```
Checking out Revision 7d1d5e7a0561458f33b01b1b25eaf52467723347 (master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 7d1d5e7a0561458f33b01b1b25eaf52467723347
Commit message: "Merge pull request #27 from daniel-beck/json"
java.lang.NullPointerException
	at io.jenkins.blueocean.autofavorite.FavoritingScmListener.onCheckout(FavoritingScmListener.java:70)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:127)
	at org.jenkinsci.plugins.workflow.libs.SCMSourceRetriever.doRetrieve(SCMSourceRetriever.java:112)
	at org.jenkinsci.plugins.workflow.libs.SCMSourceRetriever.retrieve(SCMSourceRetriever.java:84)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.retrieve(LibraryAdder.java:153)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.add(LibraryAdder.java:134)
	at org.jenkinsci.plugins.workflow.libs.LibraryDecorator$1.call(LibraryDecorator.java:125)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1065)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:133)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:127)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:557)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:518)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:290)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)

GitHub has been notified of this commit?s build result

org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: Loading libraries failed

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1085)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:133)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:127)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:557)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:518)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:290)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
Finished: FAILURE
```

jacob-keller commented:
> Unfortunately, one of the earliest jobs I checked reported a new null pointer exception. I see it in many of my jobs. Before I spend time describing the detailed steps to reporduce the problem, could you try your build of the plugin in an environment that uses a Pipeline shared library.

Yea I'll take a look, and see if I can spot it.

jacob-keller commented:
@MarkEWaite I'm pretty sure this is just a breaking change because the blueocean autofavorites plugin is searching for BuildData and expecting it to return non-null.

I suspect it's reasonably considered a bug in the FavoritingScmListener code, which needs to be null-tolerant about the missing BuildData, but previously it was "always there" so it never had issues...

This is an example of the "breaking change" I mentioned before. We no longer store BuildData actions (because they bloat the build.xml with potentially unbounded hashes), so ofcourse it's going to fail.

jacob-keller commented:
I'll have a pull request up for blueocean-autofavorites shortly to fix the null pointer exception.

jacob-keller commented:
See https://github.com/jenkinsci/blueocean-autofavorite-plugin/pull/16

jacob-keller commented:
I also added stapler/jelly code for displaying the build details now (sorry I forgot that the first time!)

and I fixed a small issue in getBySHA1 regarding failure to parse every action of a build.

jacob-keller commented:
Apparently something I changed in the last update broke a test in a weird way, hope to investigate tomorrow.

jacob-keller commented:
The failed test here doesn't fail for me locally, so I'm not exactly sure why it failed in the pipeline.

jacob-keller commented:
@MarkEWaite Is it possible to re-trigger the remote tests? They don't fail for me on the command line. I'm not sure what would be different about them on that system to cause a failure: Specifically, it seems something is timing out...

`Stacktrace
org.junit.runners.model.TestTimedOutException: test timed out after 180 seconds
at java.lang.Object.wait(Native Method)
at java.lang.Object.wait(Object.java:502)
at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:267)
at jenkins.InitReactorRunner.run(InitReactorRunner.java:44)
at jenkins.model.Jenkins.executeReactor(Jenkins.java:916)
at jenkins.model.Jenkins.<init>(Jenkins.java:815)
at hudson.model.Hudson.<init>(Hudson.java:85)
at org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:611)
at org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:384)
at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:537)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.lang.Thread.run(Thread.java:748)`

jacob-keller commented:
> @jacob-keller I merged your changes with the contents of the require-jdk-8 branch and the daniel-beck-make-matrix-optional branch and installed it into my Jenkins environment for testing.

> Unfortunately, one of the earliest jobs I checked reported a new null pointer exception. I see it in many of my jobs. Before I spend time describing the detailed steps to reporduce the problem, could you try your build of the plugin in an environment that uses a Pipeline shared library?

I locally built a pipeline library and set it to use modern SCM, and it worked for me. Note that the exception you pasted above is fixed by https://github.com/jenkinsci/blueocean-autofavorite-plugin/pull/16 which was just merged.

I do not understand why the server CI test fails still though... I can't get it to fail like that on my machines.

jacob-keller commented:
I'm pretty sure the test failures here are due to the lack of entropy on the server (SSHD fails to start in time while blocking on /dev/random and the tests timeout). I don't believe anything that remains is a problem in my code here. Is there a way to re-run the checks?

jiangty-addepar commented:
I'm excited for this to be released!

jacob-keller commented:
I've interactively tested this on a small job and things seem to be ok, but I don't know a good way to test with thousands of jobs as the history..

jiangty-addepar commented:
You could try running small jobs with a really tight cron schedule?

jacob-keller commented:
Good idea, I'll try to set my job up to do that and see what happens to the build data.

jacob-keller commented:
I've been running a job with a 1minute timer, and I'm up to around 1200 builds all without the old BuildData. So far, startup time for the job appears to be fairly stable, (about 1 second between when the pipeline starts building on the node to when the first git commands are executed according to the timestamper values).

I'll leave this running for a few days and see how bad it is when it reaches ~10,000 builds.

jacob-keller commented:
I've had this running on my system for a few weeks now, with a cron job generating a completely new (disconnected) branch in order to simulate the pathological case of many thousands of branches. I'm up to about 20k branches now.

However, with verbose logging, we can see the following 

> 00:00:00.001 Started by an SCM change
> 00:00:00.001 Started by an SCM change
> 00:00:00.001 Started by an SCM change
> 00:00:00.002 Building in workspace /var/lib/jenkins/workspace/git-build-details/many-branches
> 00:00:00.002 Using checkout strategy: Default

With verbose logging, this line is displayed just prior to retrieving build data

> 00:00:02.677 Last Built Revision: Revision 9d96b2b546eb41f51cf9b0cba44cb17366b5f436 (origin/branch/19133)

And this line is printed just after retrieving build data. With timestamping enabled, we can see that only a few seconds at most is spent, even when we consider one of the most pathological cases. Does anyone have data from the old build data for comparison? I'd like to see how this is different. I *do* know that memory is a huge issue, as is serializing data from the incredibly large build.xml data file that is generated as a result of compound copies of the builddata hashes....

> 00:00:06.212  > git rev-parse --is-inside-work-tree # timeout=10
> 00:00:06.217 Fetching changes from the remote Git repository
> 00:00:06.217  > git config remote.origin.url ssh://vagrant@localhost/home/vagrant/git/many-branches # timeout=10
> 00:00:06.222 Fetching upstream changes from ssh://vagrant@localhost/home/vagrant/git/many-branches
> 00:00:06.222  > git --version # timeout=10
> 00:00:06.226 using GIT_ASKPASS to set credentials Vagrant SSH credentials
> 00:00:06.227  > git fetch --tags --progress ssh://vagrant@localhost/home/vagrant/git/many-branches +refs/heads/*:refs/remotes/origin/*

It should also be noted that many other issues due to timing crop up with such a large number of branches, and that this code doesn't appear as a bottleneck.

jacob-keller commented:
A quick check of the build.xml files shows they no longer scale in size. It does appear that in the pathological case the log output begins to take up a very large portion (especially if verbose logging is enabled), as every job prints out thousands of builds.

It may be worth limiting the display to not show every single branch in a "see branch .. in repository" lines... or at least move that to a verbose output? It may be simply due to how changes are being picked, I'm not sure.

jacob-keller commented:
I could rebase this, if that would help, it doesn't look like the merge commit does anything incorrect, tho it looks like you restored the warning this is trying to resolve?

jacob-keller commented:
I rebased the series on top of current master :)

MarkEWaite commented:
Thanks!

MarkEWaite commented:
@basil no update on when it will be released.  The testing from @jacob-keller sounded very promising but I've not seen any reports of others that have tested it.

You could download the [latest pull request build](https://ci.jenkins.io/job/Plugins/job/git-plugin/job/PR-578/25/artifact/org/jenkins-ci/plugins/git/4.0.0-rc1685.fc5741e0cd2d/git-4.0.0-rc1685.fc5741e0cd2d.hpi), install it in your environment, confirm that it resolves your issue, and note if there are unexpected side effects in your environment.  That would help spread the work of evaluating the pull request and provide another voice declaring that the change meets the user need.

tsuna commented:
As I commented on the JIRA issue, I've been running the patch for a while and it's been working fine. Just one (small) data point though...

MarkEWaite commented:
Thanks, @tsuna !  Every user willing to test is a big help, especially with a change that has a history of attempted fixes like this one does.

basil commented:
Thanks for the instructions, @MarkEWaite. Unfortunately, I'm not willing to test this PR, since the Jenkins instances I manage are used by over a hundred engineers and I cannot afford any unnecessary issues. That having been said, I look forward to the completion of the testing for this change and for a fix to be released!

MarkEWaite commented:
I deployed the git plugin master branch 6f96e4c1892c0c10fdd3e5f0714a581b8fb321fb merged with this pull request into my test environment.  It worked quite well except for non-stop builds when using a build chooser in a Freestyle job.

Refer to the following jobs that caused non-stop builds in that environment which do not cause non-stop builds without this pull request:

- [JENKINS-33827-Default](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/JENKINS-33827-Default/config.xml)
- [JENKINS-33827-Inverse](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/JENKINS-33827-Inverse/config.xml)

The "Ancestry" variant of that case does not show non-stop builds.

I run those jobs in my [Docker environment](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/README.md)

I suspect there are tests missing from the build chooser code in the git plugin which would have shown the problem during development.  There is likely some dependency on the existence of BuildData in the jobs.  The absence of that BuildData causes the build chooser to decide that a matching branch has not been built, even though it has been built.

jacob-keller commented:
I'm glad we caught this. I will investigate. We should try to add test
cases for this as well.

On Mon, Oct 8, 2018, 9:47 AM Mark Waite <notifications@github.com> wrote:

> I deployed the git plugin master branch 6f96e4c
> <https://github.com/jenkinsci/git-plugin/commit/6f96e4c1892c0c10fdd3e5f0714a581b8fb321fb>
> merged with this pull request into my test environment. It worked quite
> well except for non-stop builds when using a build chooser in a Freestyle
> job.
>
> Refer to the following jobs that caused non-stop builds in that
> environment which do not cause non-stop builds without this pull request:
>
>    - JENKINS-33827-Default
>    <https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/JENKINS-33827-Default/config.xml>
>    - JENKINS-33827-Inverse
>    <https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/JENKINS-33827-Inverse/config.xml>
>
> The "Ancestry" variant of that case does not show non-stop builds.
>
> I run those jobs in my Docker environment
> <https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/README.md>
>
> I suspect there are tests missing from the build chooser code in the git
> plugin which would have shown the problem during development. There is
> likely some dependency on the existence of BuildData in the jobs. The
> absence of that BuildData causes the build chooser to decide that a
> matching branch has not been built, even though it has been built.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/git-plugin/pull/578#issuecomment-427903817>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA1Y3_SjApqBRZ-bizhIyL1WBhrubvh5ks5ui4GXgaJpZM4Sujxl>
> .
>

jacob-keller commented:
Ok, I think I might see the problem. So this change does impact how *much* build data we store. Previously, we kept a copy of the entire history of build data for every build, replicated once per build.

The new strategy only keeps one build details per build. So I think what might be happening here, is that your configuration seems to be only keeping the most recent 7 builds?

This means that the build data we rebuild will only contain the most 7 recent builds, which means that it may be re-triggering the older branches, because you no longer have a build that indicates it was built.

So, the problem is I don't know if we can actually resolve this.

Fundamentally, the problem with build data is that we keep a copy per build, and we continuously add builds, and we never purge it. Not only this, but old jobs which have the copy keep it in their build.xml forever.

My new strategy only works up until you want to remember builds that already ran even if you no longer keep the build data for them around.

jacob-keller commented:
Specifically the section here:

```
    <jenkins.model.BuildDiscarderProperty>
      <strategy class="hudson.tasks.LogRotator">
        <daysToKeep>-1</daysToKeep>
        <numToKeep>7</numToKeep>
        <artifactDaysToKeep>-1</artifactDaysToKeep>
        <artifactNumToKeep>-1</artifactNumToKeep>
      </strategy>
    </jenkins.model.BuildDiscarderProperty>
```

jacob-keller commented:
Essentially, the problem is that setups which include arbitrary branches (such as gerrit trigger setups) will bloat the build data because every single build is on a new branch. This means that for every job, we continuously bloat the build data and end up with thousands of builds, and each build has thousands of records for which branches did what.

Fundamentally, build data *wants* to be a per-job thing, and not a per-build thing. Unfortunately, I don't know of how we can store that information.

MarkEWaite commented:
Thanks for spending the time to investigate.  I like your description very much.  I'll change that job definition to retain enough history to assure that at least one build from each of the candidate branches will fit in the history.  That seems like a perfectly reasonable approach, and a good thing to describe in the release notes.

I think we tell people that if they want the history of all builds, they need to retain the history of all builds and not rely on the awful prior behavior of retaining the history of all builds in every build.

jacob-keller commented:
> I think we tell people that if they want the history of all builds, they need to retain the history of all builds and not rely on the awful prior behavior of retaining the history of all builds in every build.

Right. I'm glad we're on the same page here.

> That seems like a perfectly reasonable approach and a good thing to describe in the release notes.

Yea, it should be described nicely. Also note that the solution does accept using a build data that was already stored, so until the job which has build data is deleted, we will remember things up to that point.

Maybe we want to provide some sort of "cleaner" script which converts or deletes the old build data? Not sure how useful it is.

> I'll change that job definition to retain enough history to assure that at least one build from each of the candidate branches will fit in the history. 

Please let me know if this doesn't fix the issue. I am fairly sure it will, but if we do retain more jobs and still see the issue I'm inclined to think there is a bug still lurking.

> Thanks for spending the time to investigate. I like your description very much.

No problem. I've been trying to think of solutions for the build data problem for a long time, since I've got a couple Jenkins instances which are retaining huge amounts of BuildData and it causes significant slowdown and memory issues.

MarkEWaite commented:
I believe that this is ready to merge as one of the key improvements for the upcoming release of git plugin 4.0.0.  I've completed several rounds of testing successfully, have confirmed that the build chooser problem I was seeing is resolved by retaining enough jobs to remember all the branches that were chosen.

I will prepare some additional documentation for this change to be included in the release notes for git plugin 4.0.  Once that documentation is complete and on the plugin wiki page, I'll merge this pull request.

Thanks very much @jacob-keller for your excellent work on this change!  Jenkins servers all over the world should reduce their memory footprint thanks to your work.

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+        this.build = build;
+    }
+
+    public BuildDetails(Build build, String scmName) {
```
These two lesser constructors appear unused—so delete them.

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+     */
+    public String getDisplayName() {
+        if (scmName != null && !scmName.isEmpty())
+            return "Git Build Details:" + scmName;
```
Space after colon?

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+     *
+     * @param index the index, indexes less than or equal to {@code 1} will be discarded.
+     */
+    public void setIndex(Integer index) {
```
Why would you pass null here? Why not just an `int`?

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+        if (req == null) {
+            return null;
+        }
+        return req.findAncestorObject(Run.class);
```
I think you rather want to implement `RunAction2` and keep a `private transient Run<?, ?> run` field.

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * The current build.
+     */
+    public Build build;
```
Would you not want this (or its getter) to be `@Exported` too?

jglick commented on src/main/resources/hudson/plugins/git/util/BuildDetails/index_it.properties:
```diff
@@ -0,0 +1,2 @@
+Git\ Build\ Details=Dati del progetto git
+Revision=Revisione
```
Best to leave such things to translators. (Anyway the movement is towards having translations in separate repositories.)

jglick commented on src/main/java/hudson/plugins/git/util/BuildDetails.java:
```diff
@@ -0,0 +1,303 @@
+
+    public Object readResolve() {
+        if(this.remoteUrls == null)
+            this.remoteUrls = new HashSet<>();
```
Why would this be necessary in a newly introduced class? There will be no existing instances, and you can declare from the start that `remoteUrls` is `@Nonnull`.

jglick commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1177,25 +1194,28 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
+        /* Generate a BuildDetails after determining what revision to build */
+        BuildDetails buildDetails = new BuildDetails(buildData.lastBuild, getScmName(), getUserRemoteConfigs());
+
         // Track whether we're trying to add a duplicate BuildData, now that it's been updated with
```
Presumably this comment should now be edited too?

jglick commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1762,26 +1791,105 @@ public BuildData copyBuildData(Run build) {
+
             }
+
+            /* Keep digging through build history until we run out or
```
Not sure what the previous behavior was—under which circumstances the loop ran, and for how long—but any loop like this is a red flag in Jenkins as it is breaking the lazy-loading system introduced (IIRC) in 1.484. If you have a project with thousands of build records saved, reading each `build.xml` from disk and then unmarshalling its contents into Java objects can be quite expensive. (RAM usage balloons, and eventually `SoftReference`s get thrown out, at which point the _next_ such loop does it all over again, basically making the server grind to a halt.) Therefore we make every effort to avoid unbounded iteration of build records, at least in performance-critical points—i.e., any time during HTTP rendering or routine build processes.

Now if this code is only run once per historical build (or project?) and caches the relevant information along the way so that it need not loop the next time, that is OK.

jacob-keller replied:
So, the previous behavior was that each build stored its own copy of the entire build data structure. This caused significant problem because *every* build.xml contained the entire build data history, which balloons if you have unique branches per build. (such as with gerrit).

jacob-keller replied:
> Now if this code is only run once per historical build (or project?) and caches the relevant information along the way so that it need not loop the next time, that is OK.

This is definitely run once per build. I did test it using a server that generated thousands of builds, and didn't see any immediate problems. It's definitely not the most ideal solution, so better suggestions would be welcome.

I opted to use this method because it allowed us to keep feature parity with the older system and seemed to work ok.

Dropping the build data entirely is another option, but we lose the features it provides. Maybe those aren't worth it.... but I wasn't the one to make that call.

jacob-keller replied:
More comments on this implementation vs the older one.

So *every* build has to load the BuildData somehow.

In the old implementation, we deserialize it from the build.xml of the previous build, *and* we update, copy and store it in the new build.xml of the next build. If you have unique branches, then this will continue to grow for every build, plus there is no mechanism to clean it out.

In the new method, we re-build the data by checking each previous build.xml for build details. Either way the same amount of data (or less, see below) must be read in the new way as in the old way. The old way *still* loads the data into memory and thus still bloats memory and takes time to garbage collect.

The new way has a few advantages:

1) we don't copy the build data and store outdated copies for every build.mxl
2) we don't keep infinite build data that can grow arbitrarily large, because we only rebuild based on the builds which have history kept. The old method build data could keep growing unbounded when using something like gerrit refs for branches, and it would only ever clean up manually.

The only difference I can tell, is that now we read build details from multiple build.xml files, (one per historical job), vs reading the data deserialized from one build.xml. However, the sizes are going to be the same *or smaller* in the new method. The one disadvantage is that in the old world of a single branch, the build data would not grow out of bounds, so it would be trivial to keep track of, and it is definitely more efficient, since we'd have to keep building back in time.

As far as I know though, we don't generate build data except during choosing what to build, so it shouldn't be a thing called back from the webpage, but only during git operations.

I think using a separate xml file could work as well, but is an even further significant refactor. Additionally, so is simply dropping the build data entirely, as there are potentially features which depend on it. I haven't figured the scope of what we'd lose, and I'm not the plugin maintainer who could make that decision.

In my tests, I had a jenkins server which ran with several thousand jobs as the history and it didn't appear to cause significant problems.

I'm fairly confident that this method is better than build data was, even if it's not the perfect solution.

jglick commented:
> Fundamentally, build data _wants_ to be a per-job thing, and not a per-build thing. Unfortunately, I don't know of how we can store that information.

Well, if you wanted to go that route, you would most naturally just define a new `XmlFile` in the project root directory, rather than it being part of the state of a `Run`. You would have to do some work to have it clean up after deleted builds and so on.

jacob-keller commented:
> Well, if you wanted to go that route, you would most naturally just define a new XmlFile in the project root directory, rather than it being part of the state of a Run. You would have to do some work to have it clean up after deleted builds and so on.

I have no idea how to do that.

jacob-keller commented:
I have no idea how the XmlFile thing would work ,and I suspect that would take to implement, as well as verify. It would be a significant rework to either this series, or scrapping and starting from master again.

@jglick how strongly do you feel about this? what about the smaller changes?

My own testing plus the testing from @MarkEWaite  seems to indicate that this behaves well and we haven't yet noticed significant issues so far (obviously we could have missed something). Note that we're scanning this during I think SCM polling and the beginning of a build, not during web page rendering. (It actually reduces how much data is in the build xml, because we only store one revision object per build instead of copying the entire build data for every build.)

I'd like to know more thoughts given that there was previously hope to have this in 4.0 beta.

I see a few paths forward here, and I'd like to know which one we actually want to go with.

- We take this as is (or with the minor fixups you mentioned earlier)
- We decide that removing BuildData (and all the features associated with it) entirely is acceptable in a big release
- We use an XmlFile implementation which adds a separate file to the project directory
- We do nothing
- Other???

I don't know enough about the ramifications of removing BuildData entirely (which is why I invented this approach instead!)

I don't understand how to implement the XmlFile, maybe this is actually simple and I'm not understanding it very well.

I don't like the idea of doing nothing, since this is one of the significant issues with the git plugin as it stands today.

I'd like more input so that we can make a good decision.

MarkEWaite commented:
@jacob-keller I prefer we have you apply the minor fix-ups that Jesse requested.  I'm continuing my testing and have enlisted additional help from others to test as well.  I included this change in the beta-3.0 branch and delivered it as part of git client plugin 3.0.0-beta6.

jacob-keller commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -967,10 +972,22 @@ public String getGitExe(Node builtOn, EnvVars env, TaskListener listener) {
+            /* First, try BuildDetails.class */
+            List<BuildDetails> buildDetailsActions = b.getActions(BuildDetails.class);
+            for (BuildDetails details : buildDetailsActions) {
+                if (details != null && details.build != null) {
```
getBySha1 is being removed in the upcoming release, so this code was dropped.

jacob-keller commented:
Ok, I think I covered all the basic things that @jglick  suggested. If there's  other small fixup I missed please let me know :)

MarkEWaite commented:
@jacob-keller I've included your latest changes in the beta-4.0 branch.  I did not deliver a new beta release of the plugin because I didn't think those changes were especially risky.  Let me know if you think I should deliver a new beta release with those changes.

I'm continuing the testing of various parts of the git plugin 4.0 beta with this code.

jacob-keller commented:
Since you already included the bulk of changes in the earlier beta, I think that's fine. It wouldn't hurt to get more testing, but I think if we're confident then there isn't a huge reason to delay. Do you think a 4.0 beta would get a wider audience than the last beta did?

MarkEWaite commented:
It is getting a slightly larger audience than the earlier beta because @kshultzCB is testing it.  We've had others testing as well (as evidenced by [JENKINS-55387](https://issues.jenkins-ci.org/browse/JENKINS-55387?filter=14041)).

Not enough testing yet, but enough to be discovering new issues that need to be addressed.  The new issues have not been related to BuildData.  When an issue is related to BuildData, I'll post it here.

jacob-keller commented:
Ok sounds great!

jglick commented:
> @jglick how strongly do you feel about this?

Not strongly. Do not consider me an official reviewer—just trying to offer some very quick hints in case they are helpful.

>> you would most naturally just define a new `XmlFile` in the project root directory
> 
> I have no idea how to do that

You would create something along the lines of

```java
new XmlFile(build.getParent().getRootDir(), "buildData.xml")
```

and then `read` / `write` a POJO defining the per-job metadata you wished to keep, with whatever internal structure makes sense—`ArrayList`s, `HashMap`s, Jenkins specifics like `Build`, etc. Much like storing a per-build `Action` except you are responsible for explicitly loading and saving the object.

> How would this handle multiple builds running at the same time?

Well, that would be one of the tricky parts (along with cleaning up stale metadata after builds are deleted). I suppose you would most simply synchronize on the `Job` or just have a single (Jenkins-wide) static lock.

jacob-keller commented:
> > @jglick how strongly do you feel about this?
> 
> Not strongly. Do not consider me an official reviewer—just trying to offer some very quick hints in case they are helpful.
> 

Ok

> > > you would most naturally just define a new `XmlFile` in the project root directory
> > 
> > 
> > I have no idea how to do that
> 
> You would create something along the lines of
> 
> ```java
> new XmlFile(build.getParent().getRootDir(), "buildData.xml")
> ```
> 
> and then `read` / `write` a POJO defining the per-job metadata you wished to keep, with whatever internal structure makes sense—`ArrayList`s, `HashMap`s, Jenkins specifics like `Build`, etc. Much like storing a per-build `Action` except you are responsible for explicitly loading and saving the object.
> 

Ok, that makes some sense.

> > How would this handle multiple builds running at the same time?
> 
> Well, that would be one of the tricky parts (along with cleaning up stale metadata after builds are deleted). I suppose you would most simply synchronize on the `Job` or just have a single (Jenkins-wide) static lock.

The old mechanism of BuildData in each build.xml actually never cleaned itself up either. (So in some sense it's a behavior change that the new method will cleanup when old builds are deleted).

In terms of locking, Synchronizing on the Job makes sense to me, as the file is in the Job's root dir anwyays.

If I have time in the next few days I might see how difficult this is to implement.

jacob-keller commented:
For reference, I started a WIP pull request against this pull request here at https://github.com/jacob-keller/git-plugin/pull/1

Based on initial work, I think it's somewhat doable to switch to an XmlFile approach, but I'm not sure it's worth it, unless testing of the current implementation proves that loading all the old builds is a performance or memory issue.

jacob-keller commented:
Since we found that it is indeed a performance issue with many thousands of jobs, I think we should seriously investigate the XMLFile approach, where we convert BuildData to a file stored in the job directory (rather than once per build). This approach means that it will only scale linearly to the size of the jobs instead of basically the n^2 issue we have now.

I haven't had any time to figure that approach out recently, but I do think it's the best approach to resolving this issue without losing any capability.
